sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,508 @@
+  /**
+   * Utility to build up display metadata from a component and its included
+   * subcomponents.
+   */
+  public interface Builder {
+    /**
+     * Include display metadata from the specified subcomponent. For example, a {@link ParDo}
+     * transform includes display metadata from the encapsulated {@link DoFn}.
+     *
+     * @return A builder instance to continue to build in a fluent-style.
+     */
+    Builder include(HasDisplayData subComponent);
+
+    /**
+     * Register the given string display metadata. The metadata item will be registered with type
+     * {@link DisplayData.Type#STRING}, and is identified by the specified key and namespace from
+     * the current transform or component.
+     */
+    ItemBuilder add(String key, String value);
+
+    /**
+     * Register the given numeric display metadata. The metadata item will be registered with type
+     * {@link DisplayData.Type#INTEGER}, and is identified by the specified key and namespace from
+     * the current transform or component.
+     */
+    ItemBuilder add(String key, long value);
+
+    /**
+     * Register the given floating point display metadata. The metadata item will be registered with
+     * type {@link DisplayData.Type#FLOAT}, and is identified by the specified key and namespace
+     * from the current transform or component.
+     */
+    ItemBuilder add(String key, double value);
+
+    /**
+     * Register the given timestamp display metadata. The metadata item will be registered with type
+     * {@link DisplayData.Type#TIMESTAMP}, and is identified by the specified key and namespace from
+     * the current transform or component.
+     */
+    ItemBuilder add(String key, Instant value);
+
+    /**
+     * Register the given duration display metadata. The metadata item will be registered with type
+     * {@link DisplayData.Type#DURATION}, and is identified by the specified key and namespace from
+     * the current transform or component.
+     */
+    ItemBuilder add(String key, Duration value);
+
+    /**
+     * Register the given class display metadata. The metadata item will be registered with type
+     * {@link DisplayData.Type#JAVA_CLASS}, and is identified by the specified key and namespace
+     * from the current transform or component.
+     */
+    ItemBuilder add(String key, Class<?> value);
+  }
+
+  /**
+   * Utility to append optional fields to display metadata, or register additional display metadata
+   * items.
+   */
+  public interface ItemBuilder extends Builder {
+    /**
+     * Add a human-readable label to describe the previously-registered metadata field.
+     * A label is optional; if unspecified, UIs should display the metadata key to identify the
+     * display item.
+     */
+    ItemBuilder withLabel(String label);
+
+    /**
+     * Add a link URL to the previously-registered display metadata. A link URL is optional and
+     * can be provided to point the reader to additional details about the metadata.
```

Comment:
It might be helpful to add an example value for both `withLabel` and `withLinkUrl`. It helps answer things like "what constitutes a reasonable label" and "should the URL contain the http://" etc.